add integration tests for private repos

taylormadore opened the pull request:
CLOUDBLD-12547

In environments where cachito workers are configured with the credentials necessary to access private
repositories, test requests for a private repo using both https and ssh.

Tested here, with a workaround for the test_env_vars changes not being present in downstream: https://container-build-jenkins.engineering.redhat.com/job/tmadore-TEST-cachito-stage-integration-tests/5/console

And where the env is skipped, as local/cachito-test would be: https://container-build-jenkins.engineering.redhat.com/job/tmadore-TEST-cachito-stage-integration-tests/6/console

Signed-off-by: Taylor Madore <tmadore@redhat.com>

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] New code has type annotations
- n/a OpenAPI schema is updated (if applicable)
- n/a DB schema change has corresponding DB migration (if applicable)
- n/a README updated (if worker configuration changed, or if applicable)
- n/a Draft release notes are updated before merging

chmeliik commented on tests/integration/test_private_repos.py:
```diff
@@ -0,0 +1,48 @@
+    client.download_and_extract_archive(completed_response.id, tmpdir)
+    source_path = tmpdir.join(f"download_{str(completed_response.id)}")
+    repo = Repo(os.path.join(source_path, "app"))
+    assert repo.head.commit.hexsha == env_data["ref"]
```
additional idea: check that `repo.git.diff()` is empty after checking the expected ref => we're basically checking all the expected files

taylormadore replied:
Nice! Will add

ejegrova commented on test_env_vars.yaml:
```diff
@@ -244,3 +244,8 @@ various_packages:
     repo: https://github.com/cachito-testing/cachito-gomod-test
     ref: 1827221d787cbd1e979b339cfbbf59728eddf0d4
     dependencies_count: 48
+# Test data for the private repos test
+# The test will only be performed if $JOB_NAME includes one of the following
+private_repo_test_envs:
+  - cachito-prod
+  - cachito-stage
```
Can we have this data elsewhere please? I believe [tests/integration/test_data/private_repo_packages.yaml](https://github.com/containerbuildsystem/cachito/pull/779/files#diff-c0f35f02e13615fcf7b7925d912b94e8299aaf78c70b071d0591f12438f5ee8d) should be fine for that. We wanted to move all specific test data from this file because we have to change it also downstream every time we make a change here.

taylormadore replied:
Moved to the private repo test data file :+1:

ejegrova commented on tests/integration/test_data/private_repo_packages.yaml:
```diff
@@ -0,0 +1,6 @@
+https:
```
nitpick: It might be nice to keep the same structure as in the other test_data files and add name of the test as the first key.

taylormadore replied:
Just to clarify, do you think I should change the https one to just "private_repo"?

ejegrova replied:
no, sorry, that's not what I meant. Check out for example: https://github.com/containerbuildsystem/cachito/blob/master/tests/integration/test_data/cached_dependencies.yaml#L2
We have one key with name of the test `cached_package` and all the appropriate variables for that test are nested under that key. That way when there's another test we need to add for caching, we just use the same file for test data.

chmeliik replied:
But `private_repo_https` and `private_repo_ssh` are two different test cases, right? The structure makes sense to me

ejegrova replied:
oh right, it's fine then

chmeliik commented on tests/integration/test_private_repos.py:
```diff
@@ -0,0 +1,63 @@
+    * Check that the request completes successfully
+    * Check that no packages are identified in response
+    * Check that no dependencies are identified in response
+    * Check that the source tarball includes the application source code
```
Thanks for adding the docstring :tada: 

Could we also mention how we check the application source code here, and why we have to do it differently than the other tests?

chmeliik commented:
Do we care about type annotations in test code? Probably not that much, right?

chmeliik reviewed (approved):
LGTM :+1:

taylormadore commented:
> Do we care about type annotations in test code? Probably not that much, right?

It looks like we don't run mypy on the tests, but I added the annotations anyway :+1: 
In the process of doing so, I saw that the pytest tmpdir fixture [is deprecated-ish](https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html#the-tmpdir-and-tmpdir-factory-fixtures), so I switched it to use tmp_path instead.

chmeliik reviewed (approved):
> In the process of doing so, I saw that the pytest tmpdir fixture [is deprecated-ish](https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html#the-tmpdir-and-tmpdir-factory-fixtures), so I switched it to use tmp_path instead.

Yay for pathlib! :tada:
